File: python/combp/option.py
"""
オプショナルコンビネータ。
"""

from __future__ import annotations

from .combinator_types import Combinator, CombinatorResult, ok


def option[C, T](combinator: Combinator[C, T]) -> Combinator[C, T | None]:
    """
    コンビネータを試し、失敗した場合はエラーなしで None を返す。

    成功した場合は結果をそのまま返す。
    正規表現の ``?`` に相当する。コンテキストは失敗時には進まない。

    型パラメータ:
        C: コンテキスト型。
        T: 内側のコンビネータの結果型。

    Args:
        combinator: 省略可能な対象のコンビネータ。

    Returns:
        成功時に T 型の値、失敗時に None を返すコンビネータ。
        いずれの場合も ok=True となる。

    Example::

        c = option(same(","))
        r = c(Ctx("abc"))
        # r.ok == True, r.value is None
    """

    def _option(context: C) -> CombinatorResult[C, T | None]:
        result = combinator(context)
        if result.ok:
            return result  # type: ignore[return-value]
        return ok(context, None)

    return _option

File: go/option.go
package combp

// Option はコンビネータを試し、失敗した場合はエラーなしで nil を返す。
//
// 成功した場合は *T（値へのポインタ）を返す。正規表現の ? に相当する。
// 失敗時にコンテキストは進まない。いずれの場合も結果は成功となる。
//
// 型パラメータ:
//   - C: コンテキスト型。
//   - T: 内側のコンビネータの結果型。
//
// 引数:
//   - c: 省略可能な対象のコンビネータ。
//
// 戻り値: 成功時に *T、失敗時に (*T)(nil) を返すコンビネータ。
func Option[C any, T any](c Combinator[C, T]) Combinator[C, *T] {
	return func(ctx C) Result[C, *T] {
		r := c(ctx)
		if r.OK {
			v := r.Value
			return OK(r.Context, &v)
		}
		return OK(ctx, (*T)(nil))
	}
}
